// token/token.js
/*
	Author: Stéphane Bascher
	Freebox token
	Date: May--2015 - Version: 1.0 - Creation of the module
*/

var request = require('request');

// Init js
var token = module.exports = function (opts) {
	//Dirty hack
	tokenobj = this;

	if (!(this instanceof token)) {
		return new token();
	}
	
	opts = opts || {};
	this.debug = 'false';  
	
	this.app = this.app || {
		status        : 'granted',
		logged_in     : false,
		challenge     : '',
		password      : '',
		session_token : '',
		permissions   : {}
	};
	
	// Save action
	this.PlayerOn = function (token, id, version, callback) {this.sessionRequest(token, id, version, function (state) { 
								if (state == true)
									tokenobj.IsOn(tokenobj, function(state) { 
										callback(state);
									});
								else
									callback(state);
							})};								
}	



token.prototype.IsOn = function(client,callback) {
	
	var options = {
		url    : "http://mafreebox.freebox.fr/api/v3/airmedia/receivers/Freebox%20Player/",
		headers : {
				'X-Fbx-App-Auth' : client.app.session_token
			}, 
		method : 'POST',
		json   : {
			  'action': 'stop',
			  'media_type' : 'video'
			  },
		encode : 'utf-8'
	};

	request(options, function (error, response, body) {
		if (!error && response.statusCode == 200) {
			callback(body.success);
		} else 
			callback(-1);
	});
}




token.prototype.sessionRequest = function (token, id, version, callback) {
	var client = this;	
	
	var crypto  = require('crypto');
	
	request('http://mafreebox.freebox.fr/api/v1/login/', function (error, response, body) {

		if (!error && response.statusCode == 200) {
			body = JSON.parse(body);
			client.app.logged_in = body.result.logged_in; 
			client.app.challenge = body.result.challenge; 
			client.app.password = crypto.createHmac('sha1', token).update(client.app.challenge).digest('hex'); 
			
			if (client.debug == 'true') {
				info("logged in:",client.app.logged_in ); 
				info("update challenge:",body.result.challenge); 
				info("password:",client.app.password);
			}

			if (!client.app.logged_in){
				var options = {
					url    : 'http://mafreebox.freebox.fr/api/v1/login/session/',
					method : 'POST',
					json   : {
					   "app_id"      : id,
					   "app_version" : version,
					   "password"    : client.app.password,
					},
					encode : 'utf-8'
				};

				request(options, function (error, response, body) {
					if ( !error && (response.statusCode == 200 || response.statusCode == 403) ) {
						client.app.challenge = body.result.challenge; 
						
						if (client.debug == 'true') info("login status:",client.app.logged_in );
						
						if (response.statusCode == 200) { 
							client.app.session_token = body.result.session_token; 
							client.app.logged_in   = true; 
							client.app.permissions = body.result.permissions;

							if (client.debug == 'true') {
								info("session token:",client.app.session_token); 
								info("logged in:",client.app.logged_in); 
								info("permissions:",client.app.permissions); 
							}
							callback(true);
						}
						else if(response.statusCode == 403) { 
							client.app.logged_in = false; 
							error(body.msg + ' : ' + body.error_code);
							callback(-1);
						}
					} else {
						callback(-1);
					}
				});
			} else
				callback(true);
		} else {
			callback(-1);
		}
	});
}
